exonum/src/messages2/authorization.rs

```diff
@@ -0,0 +1,145 @@
+    ) -> SignedMessage {
+        let mut buffer = Vec::new();
+        buffer.extend_from_slice(author.as_ref());
+        buffer.push(cls);
+        buffer.push(tag);
+        buffer.extend_from_slice(value.as_ref());
+        let signature = Self::sign(&buffer, secret_key).expect("Couldn't form signature");
+        buffer.extend_from_slice(signature.as_ref());
+        SignedMessage { raw: buffer }
+    }
+
+    /// Creates `SignedMessage` wrapper from the raw buffer.
+    /// Checks binary format and signature.
+    pub fn verify_buffer(buffer: Vec<u8>) -> Result<Self, Error> {
+        if buffer.len() <= EMPTY_SIGNED_MESSAGE_SIZE {
+            bail!("Message too short message_len = {}", buffer.len())
+        }
+        let signed = SignedMessage { raw: buffer };
+
+        let pk = signed.author();
+        let signature = signed.signature();
+
+        Self::verify(signed.data_without_signature(), &signature, &pk)?;
+
+        Ok(signed)
+    }
+
+    fn data_without_signature(&self) -> &[u8]{
+        debug_assert!(self.raw.len() > EMPTY_SIGNED_MESSAGE_SIZE);
+        let sign_idx = self.raw.len() - SIGNATURE_LENGTH;
+        &self.raw[0..sign_idx]
+    }
+
+    #[cfg(test)]
+    pub(crate) fn unchecked_from_vec(buffer: Vec<u8>) -> Self {
+        SignedMessage { raw: buffer }
+    }
+
+    #[cfg(not(test))]
+    pub(in messages) fn unchecked_from_vec(buffer: Vec<u8>) -> Self {
+        SignedMessage { raw: buffer }
+    }
+
+    pub(in messages) fn author(&self) -> PublicKey {
+        PublicKey::from_slice(&self.raw[0..PUBLIC_KEY_LENGTH]).expect("Couldn't read PublicKey")
+    }
+
+    pub(in messages) fn message_class(&self) -> u8 {
+        self.raw[PUBLIC_KEY_LENGTH]
+    }
+
+    pub(in messages) fn message_type(&self) -> u8 {
+        self.raw[PUBLIC_KEY_LENGTH + 1]
+    }
+
+    pub(in messages) fn payload(&self) -> &[u8] {
+        let sign_idx = self.raw.len() - SIGNATURE_LENGTH;
+        &self.raw[PUBLIC_KEY_LENGTH + 2..sign_idx]
+    }
+
+    pub(in messages) fn signature(&self) -> Signature {
+        let sign_idx = self.raw.len() - SIGNATURE_LENGTH;
+        Signature::from_slice(&self.raw[sign_idx..]).expect("Couldn't read signature")
+    }
+
+    /// Return byte array representation of internal data.
+    pub fn raw(&self) -> &[u8] {
+        &self.raw
+    }
+
+    pub fn hash(&self) -> Hash {
+        hash(&self.raw)
+    }
+
+    fn sign(full_buffer: &[u8], secret_key: &SecretKey) -> Result<Signature, Error> {
+        let signature = crypto::sign(&full_buffer, secret_key);
+        Ok(signature)
+    }
+
+    fn verify(
+        full_buffer: &[u8],
+        signature: &Signature,
+        public_key: &PublicKey,
+    ) -> Result<(), Error> {
+        if !crypto::verify(signature, &full_buffer, &public_key) {
+            bail!("Can't verify message.");
+        }
+        Ok(())
+    }
+}
+
+impl ToHex for SignedMessage {
+    fn write_hex<W: ::std::fmt::Write>(&self, w: &mut W) -> ::std::fmt::Result {
```

Comment:
Please avoid `::` at the beginning

Reply:
Thank you for your review, but could you clarify why i should avoid it? Neither https://github.com/exonum/exonum/blob/master/CONTRIBUTING.md
nor rustfmt code stype didn't tell anything about this.


And i prefer to keep uses as local as possible.

Reply:
@vldm 
Check `CONTRIBUTING.md` again. 🙃